lucene/analysis/common/src/java/org/apache/lucene/analysis/miscellaneous/ConcatenateGraphFilter.java

```diff
@@ -119,27 +147,19 @@ public boolean incrementToken() throws IOException {
 
   @Override
   public void end() throws IOException {
-    super.end();
-    if (finiteStrings == null) {
-      inputTokenStream.end();
-    }
-  }
-
-  @Override
-  public void close() throws IOException {
-    if (finiteStrings == null) {
-      inputTokenStream.close();
-    }
+    restoreState(endState);
   }
 
+  //nocommit move method to before incrementToken
   @Override
   public void reset() throws IOException {
-    super.reset();
-    if (hasAttribute(CharTermAttribute.class)) {
-      // we only create this if we really need it to safe the UTF-8 to UTF-16 conversion
-      charTermAttribute = getAttribute(CharTermAttribute.class);
```

Comment:
The automaton should be created lazily on the first call to incrementToken like described in the comment above ;).

Reply:
I understand what you are asking me to do (to put back) but don't understand _why_.  What comment above do you refer to?  Where your feedback is?  That comment doesn't reference that; or at least it's very non-obvious to me if there's a relationship (and we should add comments explaining).

Reply:
I am referring to this comment: 
````
//nocommit move method to before incrementToken
````
but maybe I misunderstood what you meant by move. Let me ask you this differently, why do you need to build the automaton on reset ? Unless there is a good reason to do it we shouldn't change the behavior here which is to build the automaton on the first call to incrementToken.

Reply:
Oh *that* comment.  By that I mean move the _method_ to follow a natural sequence.  I like classes to have methods that are organized sensibly.  When one uses a TokenStream, reset() should come before incrementToken() !  -- then end(), then close(), and any utility methods should come after callers.

I moved the initialization to reset simply because it felt natural this way rather than incrementToken where it's lazy.  That's all.  That said... I hinted at a TestRandomChains failure about lifecycle needing to throw an IllegalStateException or something like that (pertaining to missing close()), and I think making this back to lazy will "fix" that.